### PR TITLE
23 add tests tws@main

### PR DESCRIPTION
**Pull request by BLAZEWIM**
# Pull Request

Fixes #[23](https://github.com/insightsengineering/teal.widgets/issues/23)

News.Rd & version number not changed (thought it is not needed for a test, even with new package introduced?)

shinytest2 already added [here](https://code.roche.com/nest/automation/NEST_releases/-/blob/main/renv/nest_environments/bee_4.1.1_devel.lock)

NOTE: github Code Coverage Summary does not count shinytest2 build from init/ folder as "covered".

NOTE: currently shinytest2 will not work on automation side (docker), this puts this whole idea in question.

Open to ideas.

### CODE REVIEW

File: inst/tws/tests/testthat/test-shinytest2.R [hunk 1]
```diff
@@ -0,0 +1,24 @@
+library(shinytest2)
+
+test_that("{shinytest2} recording: tws_screenshot", {
+  app <- AppDriver$new(
+    variant = platform_variant(), name = "tws_screenshot",
+    height = 823, width = 1459
+  )
+  app$expect_screenshot()
```
Comment (asbates) [inline inst/tws/tests/testthat/test-shinytest2.R:8]: I would not recommend screenshot testing. In my experience it's too finnicky and has more cost than benefit.

For example, when I ran the tests the screenshots were flagged as changed. I'm on the same `platform_variant()` too. Looking at the images, it was the tiniest difference, something a human may not even if it wasn't highlighted.

File: inst/tws/tests/testthat/test-shinytest2.R [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+})
+
+test_that("{shinytest2} recording: tws_download", {
+  app <- AppDriver$new(
+    variant = platform_variant(), name = "tws_download", height = 823,
+    width = 1459
+  )
+  app$set_window_size(width = 1459, height = 766)
```
Comment (asbates) [inline inst/tws/tests/testthat/test-shinytest2.R:16]: I think we should add some snapshot tests, recording the values of inputs and outputs. Here's an example:

```
test_that("snapshotting inputs", {

  app <- AppDriver$new(name = 'blah')

  app$expect_values(
    input = c(
      'table_with_settings-downbutton-file_format',
      'table_with_settings-downbutton-lpp',
      'table_with_settings-downbutton-pagination_switch'
    )
  )

  file_name <- app$get_value(input = 'table_with_settings-downbutton-file_name')
  file_name <- sub('_\\d{6}$', '', file_name) # remove time
  expected_file_name <- paste0("table_", strftime(Sys.Date(), format = "%Y%m%d"))
  testthat::expect_equal(file_name, expected_file_name)

  app$expect_text("table")
  app$stop()

})
```

Try adding this and looking at the snapshot files ('.json' and '.md'). Snapshotting the values like this is more robust than screenshots. And you can do any manipulation needed on the object (e.g. file name).

**Review by asbates — COMMENTED** [review]
This is a good start and gets the infrastructure set up which is often half the battle.

The main comment I have is that I recommend snapshot tests (saving values) over screenshot tests (saving images).

I know there were some issues with Shiny assigning random ids to some inputs. As an example of how this can be fixed, add an `inputId = ns("blah")` to the [`shinyWidgets::dropdownButton`](https://github.com/insightsengineering/teal.widgets/blob/2b3a0003a12009af36f25e325b1add9e4cb8e556/R/table_with_settings.R#L120). The id will then be fixed, allowing for consistent access in tests.

File: inst/tws/tests/testthat/test-shinytest2.R (hunk 1 above)
Comment (BLAZEWIM) [reply inst/tws/tests/testthat/test-shinytest2.R]: So this cannot happened. I though it is too good to be true, to have an image identical up to every pixel

**Comment by pawelru** [comment]
Hi @BLAZEWIM if it is a work-in-progress please kindly convert to a draft PR. This would secure against an accidental merge (it had happened in the past :( )

**Comment by BLAZEWIM** [comment]
> Hi @BLAZEWIM if it is a work-in-progress please kindly convert to a draft PR. This would secure against an accidental merge (it had happened in the past :( )

Ok, it is ready for review, but no shinytest2 tests will be introduced in this cycle, so I conclude that this should be turned into draft.

**Comment by BLAZEWIM** [comment]
NOTE to future self: after consideration, it is debatable whether we should add inputId = ns() to code, as it overrides the default behavior of the download. Try to replac with other sollution.

**Review by asbates — APPROVED** [review]
Looks good!

**Review by chlebowa — APPROVED** [review]

**Pull request merged**